### ISSUE TITLE
[strict-template-expressions] [no-unsafe-member-access] False positive for imported unions

### ISSUE THREAD

**Issue by limitedmage** [typescript-eslint/typescript-eslint#2723]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/restrict-template-expressions": ["error", { "allowNumber": true, "allowBoolean": true }],
    "@typescript-eslint/no-unsafe-member-access": "warn",
  }
}
```

```TS
/// file1.ts

export type Thing = { type: 'one'; data: string } | { type: 'two'; data: number }

/// file2.ts
import { Thing } from 'file1.ts'

export function convertThingToString(event: Thing): string {
    return `event: ${event.type}\ndata: ${JSON.stringify(event.data)}\n\n`
}
```

**Expected Result**

Everything is valid

**Actual Result**
Error:
```
Invalid type "any" of template literal expression. (eslint@typescript-eslint/restrict-template-expressions)
```

Warning:
```
Unsafe member access .type on an any value. (eslint@typescript-eslint/no-unsafe-member-access)
```

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.5.0` |
| `@typescript-eslint/parser`        | `4.5.0` |
| `TypeScript`                       | `4.0.3` |
| `ESLint`                           | `7.8.1` |
| `node`                             | `14.7.0` |

**Comment by bradzacher** [comment]
If you have false positives with imported values, then this generally indicates a configuration error.
The `any` isn't a true any, it's an `any` because the instance of TypeScript being used by the tooling cannot figure out what file you're referencing.

Could you please create a repro repo?

**Comment by limitedmage** [comment]
I'll try to work on a repro repo. In the meantime, is there any way to figure out why the TypeScript configuration may be wrong for ESLint but correct for TypeScript compilation?

**Comment by bradzacher** [comment]
Hard to say without more information

**Comment by limitedmage** [comment]
Looks like the issue might be related to using project references, which does not seem to be fully supported based on #2094. I massaged the `.tsconfig` file to remove the `references` line and then re-add it, and the warnings and errors disappeared.

**Comment by bradzacher** [comment]
If you're just seeing errors in the editor - then yes, this can happen, as we don't support project references yet.

Did the issues persist after restarting your IDE?

**Comment by limitedmage** [comment]
Yes, the issue did persist after restarting the IDE and I even saw it when running ESLint from the command line; it didn't go away until I edited the tsconfig file. 

I'll reopen this if I can find a consistent, isolated repro. Thanks for your help!
